### PR TITLE
  PLNSRVCE-547 Ability to configure JVM builds 

### PR DESCRIPTION
**Pull request by stuartwdouglas**
The presence of the "jvm-build-config" ConfigMap triggers the
environment setup. This config map is used to configure user related
settings in the namespace.

In addition there is the jvm-build-service/jvm-build-system-config that
configures system wide config such as builder images and other system
settings.

Presently this will:
- Create a new cache instance in the local workspace if the `jvm-build-config` configmap is present in the namespace
- Configure this cache from the map
- Configure system wide images from a system config map
- Configure builder images from a system config map

This also resolves the following:

[PLNSRVCE-465](https://issues.redhat.com//browse/PLNSRVCE-465) Java Workspace Setup
[PLNSRVCE-547](https://issues.redhat.com//browse/PLNSRVCE-547) Allow users to configure per namespace recipe repositories

Infra deployments PR: https://github.com/redhat-appstudio/infra-deployments/pull/537
Build Definitions PR: https://github.com/redhat-appstudio/build-definitions/pull/179

### CODE REVIEW

File: pkg/reconciler/configmap/configmap.go
Comment (stuartwdouglas) [reply pkg/reconciler/configmap/configmap.go:101]: We will need some form of configuration here. Big workspaces may need caches with more resources.

File: pkg/reconciler/configmap/configmap.go
Comment (stuartwdouglas) [reply pkg/reconciler/configmap/configmap.go:107]: I was thinking we could just mount the configmap in the sidecar pod and have it read this configuration directly, however for this particular bit of config I think we just want to use ${context.taskRun.namespace} to get the namespace.

File: pkg/reconciler/configmap/configmap.go
Comment (gabemontero) [reply pkg/reconciler/configmap/configmap.go:107]: yep .... and since my comment we had the cabal discussion with @sbose78 and the rest of the 3 pipeline service teams and reached agreement that there is a broader concern around setting up workspaces/namespaces with more quota based on request and additional payment, not just jvm build service

that said, in the interim, I think you should define keys in the configmap for these, where the values you've picked now could be the default if those key are missing.

File: pkg/reconciler/configmap/configmap.go
Comment (gabemontero) [reply pkg/reconciler/configmap/configmap.go:113]: +1 on both those ideas @stuartwdouglas

File: pkg/reconciler/configmap/configmap.go
Comment (stuartwdouglas) [inline pkg/reconciler/configmap/configmap.go:57]: @gabemontero can I get your opinion on this bit? Basically the builder images are needed in the dependencybuild reconciler, and it seems really wasteful to have to explcitiy grab the config map there every time we do a build. 

Instead I just have this shared BuilderImageConfig object and access it under a lock.

File: .ci/common-ci-settings-and-functions.sh
Comment (stuartwdouglas) [inline .ci/common-ci-settings-and-functions.sh:59]: @gabemontero where are these images actually built? I will need to build the builder images now as well but I can't seem to find where this build+env var setting happens.

Comment (psturc) [reply .ci/common-ci-settings-and-functions.sh:59]: @stuartwdouglas [here's the openshift-ci job config file](https://github.com/openshift/release/blob/master/ci-operator/config/redhat-appstudio/jvm-build-service/redhat-appstudio-jvm-build-service-main.yaml) where they need to be added

you can find more info in the [readme](https://github.com/redhat-appstudio/jvm-build-service/blob/main/.ci/README.md)

Comment (gabemontero) [reply .ci/common-ci-settings-and-functions.sh:59]: @stuartwdouglas this is what we opened https://issues.redhat.com/browse/PLNSRVCE-517 after our exchange start started with my comment https://github.com/redhat-appstudio/jvm-build-service/pull/106#pullrequestreview-1042028142
and ended with your comment https://github.com/redhat-appstudio/jvm-build-service/pull/106#issuecomment-1192064239

Comment (gabemontero) [reply .ci/common-ci-settings-and-functions.sh:59]: I tried the capture all the relevant steps in my update to the description in but of course we can talk, divide and conquer as needed if you are ready to take on https://issues.redhat.com/browse/PLNSRVCE-517

@Michkov FYI

File: pkg/reconciler/configmap/configmap.go
Comment (gabemontero) [reply pkg/reconciler/configmap/configmap.go:63]: yeah since we are reconciling on the config map it should be in the watch cache and hence grabbing it each time using the controller runtime client is suppose to be a "local fetch / in memory only / cache hit" with no RPC.

general rule of thumb we don't want to synchronize across events like this

File: .ci/common-ci-settings-and-functions.sh
Comment (stuartwdouglas) [reply .ci/common-ci-settings-and-functions.sh:59]: https://github.com/openshift/release/pull/30919

File: pkg/reconciler/configmap/configmap.go
Comment (gabemontero) [reply pkg/reconciler/configmap/configmap.go:139]: wanted to circle back to this thread @stuartwdouglas .... I'm not seeing anything in https://github.com/redhat-appstudio/build-definitions/pull/179/files or https://github.com/redhat-appstudio/jvm-build-service/blob/c1032fe8ca4f0b0c95d16b34361ace5df649b321/pkg/reconciler/dependencybuild/buildrecipeyaml.go around mounting in the config map

and I see you deleted the env var QUARKUS_REST_CLIENT_CACHE_SERVICE_URL I mentioned with https://github.com/redhat-appstudio/jvm-build-service/pull/105/files#r923400881

is it still a to-do (perhaps in that ^^ build def PR) to add the config map mount?  or are you going in a different direction?

thanks

**Review by gabemontero — CHANGES_REQUESTED** [review]
ok @stuartwdouglas I started looking at this PR on Monday Aug 1 after the stand up meeting we were both on and see I have a pending review from last Friday I never hit submit review on

sign - too much multi tasking for me on Friday

I've cleaned up the review up a bit and and hitting submit review now.  I'll then take an additional pass in case there was an update you made today that I need to see.

File: .ci/common-ci-settings-and-functions.sh
Comment (gabemontero) [reply .ci/common-ci-settings-and-functions.sh:57]: update for anyone coming into this thread mid stream ... https://github.com/openshift/release/pull/30919 has merged

after these changes here, next step is to add the requisite updates in infra-deps (but correct me if I'm wrong @stuartwdouglas ) to consume them.

File: pkg/reconciler/configmap/configmap.go
Comment (gabemontero) [reply pkg/reconciler/configmap/configmap.go:81]: fyi we have agreed that ^^ is beyond the scope of this PR

(though I do want to start pursuing the concept at least with @sbose78 when he returns from PTO)

File: pkg/reconciler/configmap/configmap.go
Comment (gabemontero) [inline pkg/reconciler/configmap/configmap.go:166]: I suppose my question from https://github.com/redhat-appstudio/jvm-build-service/pull/105/files#r932615502 applies here too ... need a mem refresh from one of the voice to voice chats we had last week.

File: pkg/reconciler/configmap/configmap.go
Comment (gabemontero) [inline pkg/reconciler/configmap/configmap.go:240]: Preface: I understand the need for this change here with what this PR is accomplishing.

That said, a controller adding roles and rolebindings (i.e. manipulate permissions) is atypical (and can in fact be frowned upon).

Now, needing anyuid is a function of the base images we are leveraging during our builds not giving us write access to content we need.

We are not the only item under the appstudio umbrella fighting this (the buildah task is another one, where it has a path around user workspaces for eventually solving this (though that is not there in openshift 4.10 IIRC so that I think is a big change).

But I have had brief conversations with Adam Kaplan about "long term" not wanting any appstudio components to leverage anyuid.

Bottom line, *minimally* we need a big TODO comment here about we know we need to see about create images that do not need anyuid, where that is complex enough that if we do not already have a Jira story that tracks this work, we should either update an existing story, or create a new one.

@Michkov FYI

**Comment by gabemontero** [comment]
OK done with PR review @stuartwdouglas 

e2e failure ` .ci/common-ci-settings-and-functions.sh: line 57: JVM_BUILD_SERVICE_JDK8_BUILDER_IMAGE: unbound variable ` should be addressed with the openshift/release merge I believe

/retest

**Comment by gabemontero** [comment]
bad luck with app studio bootstrap:

```
         1.6593693589970229e+09	ERROR	controller	timed out waiting for taskrun CRD to be created	{"error": "configmaps \"jvm-build-system-config\" not found"}
        github.com/redhat-appstudio/jvm-build-service/pkg/controller.NewManager
        	/opt/app-root/src/pkg/controller/controller.go:118
        main.main
        	/opt/app-root/src/cmd/controller/main.go:39
        runtime.main
        	/usr/lib/golang/src/runtime/proc.go:255
        1.6593693589970791e+09	ERROR	main	unable to start manager	{"error": "configmaps \"jvm-build-system-config\" not found"}
        main.main
        	/opt/app-root/src/cmd/controller/main.go:48
        runtime.main
        	/usr/lib/golang/src/runtime/proc.go:255
```

/retest

**Comment by gabemontero** [comment]
same error as before 

```
         1.659372107494306e+09	ERROR	controller	timed out waiting for taskrun CRD to be created	{"error": "configmaps \"jvm-build-system-config\" not found"}
        github.com/redhat-appstudio/jvm-build-service/pkg/controller.NewManager
        	/opt/app-root/src/pkg/controller/controller.go:118
        main.main
        	/opt/app-root/src/cmd/controller/main.go:39
        runtime.main
        	/usr/lib/golang/src/runtime/proc.go:255
        1.6593721074943602e+09	ERROR	main	unable to start manager	{"error": "configmaps \"jvm-build-system-config\" not found"}
        main.main
        	/opt/app-root/src/cmd/controller/main.go:48
        runtime.main
        	/usr/lib/golang/src/runtime/proc.go:255 
``` 

possibly related to my comment https://github.com/redhat-appstudio/jvm-build-service/pull/105/files#r932612117 something must be amiss with that addition, especially in light that CRD check is already in main branch at https://github.com/redhat-appstudio/jvm-build-service/blob/7d2f00b2bb9ed3de167a76e8533d03921518026d/pkg/controller/controller.go#L35-L46

I'll start comparing files in a few

File: pkg/controller/controller.go
Comment (gabemontero) [reply pkg/controller/controller.go:54]: OK I looked at the entire file and it became obvious.  the CRD check is duplicated @stuartwdouglas 

https://github.com/redhat-appstudio/jvm-build-service/blob/c1032fe8ca4f0b0c95d16b34361ace5df649b321/pkg/controller/controller.go#L40-L64

Remove this duplicate chunk

**Comment by gabemontero** [comment]
duh - I need better glasses:

`        1.659372107494306e+09	ERROR	controller	timed out waiting for taskrun CRD to be created	{"error": "configmaps \"jvm-build-system-config\" not found"}`

it is the new code looking for the system config map
though that message should be updated to not say taskrun CRD 

(that said the duplication of the CRD check needs to be cleaned up also)

this PR's test needs the PRs @stuartwdouglas noted in the description (minimally https://github.com/redhat-appstudio/infra-deployments/pull/537)

sorry for the noise @stuartwdouglas .... I'll clean up some noisy comments

File: pkg/controller/controller.go
Comment (gabemontero) [inline pkg/controller/controller.go:118]:
```suggestion
		controllerLog.Error(err, "timed out waiting for system config map to be properly intialized")
```

File: pkg/reconciler/configmap/configmap.go
Comment (stuartwdouglas) [reply pkg/reconciler/configmap/configmap.go:125]: We definitly want it to be system level configurable, probably also individually configurable (assuming we have some kind of per namespace quota, so a user can't just increment this to get more total memory).

File: pkg/reconciler/configmap/configmap.go
Comment (stuartwdouglas) [reply pkg/reconciler/configmap/configmap.go:139]: We don't actually need a config map for this, we can handle the per workspace cache URI via standard tekton config expansion. If we do want to configure the sidecar we will need to mount the config map but I want to limit the blast radius of this PR so IMHO that can come later.

File: pkg/reconciler/configmap/configmap.go
Comment (stuartwdouglas) [reply pkg/reconciler/configmap/configmap.go]: I have changed this to just get the config map in the db reconciler

File: .ci/common-ci-settings-and-functions.sh
Comment (stuartwdouglas) [reply .ci/common-ci-settings-and-functions.sh:57]: https://github.com/redhat-appstudio/infra-deployments/pull/537 configures the config map to reference them in infra deployments.

File: deploy/base/config.yaml
Comment (stuartwdouglas) [reply deploy/base/config.yaml:7]: At the moment the user would create it to enable rebuilds, in a similar manner to how they need to configure the hacbs bundle. 

There are no permissions on this yet so anyone can enable this, but I don't think this is a big deal at this stage.

File: pkg/controller/controller.go
Comment (stuartwdouglas) [reply pkg/controller/controller.go:54]: oops

File: pkg/reconciler/configmap/configmap.go
Comment (stuartwdouglas) [reply pkg/reconciler/configmap/configmap.go:240]: This PR actually changes the base image to the ones built in openshift, so I can probably make that image change as part of this PR. 

I will need to do some local testing.

File: pkg/reconciler/configmap/configmap.go
Comment (gabemontero) [reply pkg/reconciler/configmap/configmap.go:123]: we can assume per namespace quota

and I'm good with staging system level vs. individual configuration separately (with system first) if that seems like the correct path to you @stuartwdouglas 

do you want to include  in this PR or a follow up?

File: pkg/reconciler/configmap/configmap.go
Comment (gabemontero) [reply pkg/reconciler/configmap/configmap.go:137]: ah right .... and that explain why you deleted the env var ... ok yeah makes sense

I'm totally fine with deferring configurable side car / config map mount outside this PR ... might even be something we defer indefinitely - we'll see.

File: pkg/reconciler/configmap/configmap.go
Comment (gabemontero) [reply pkg/reconciler/configmap/configmap.go]: yep in conjunction with https://github.com/redhat-appstudio/jvm-build-service/pull/105/files#diff-243ebed2765f75e6a54f57167212fefb08c3b2a85967ad2acbc0eb78919019c1R84 looks good  / thanks

marking resolved

File: .ci/common-ci-settings-and-functions.sh
Comment (gabemontero) [reply .ci/common-ci-settings-and-functions.sh:57]: yep thanks for that more specif ref @stuartwdouglas 

marking resolved

File: deploy/base/config.yaml
Comment (gabemontero) [reply deploy/base/config.yaml:7]: ok thanks for the confirm

and yeah the permission etc. we agreed is a follow up item, where my idea at https://github.com/redhat-appstudio/jvm-build-service/pull/105#discussion_r923391438 is the first attempt at coming up with the approach for that future implementation

marking as resolved

File: pkg/controller/controller.go
Comment (gabemontero) [reply pkg/controller/controller.go]: looks good now marking resolved

File: pkg/reconciler/configmap/configmap.go
Comment (gabemontero) [reply pkg/reconciler/configmap/configmap.go]: ok cool - we'll play it by ear.

initial thought looking for example (but not limited to) we would tweak https://github.com/redhat-appstudio/jvm-build-service/blob/main/Dockerfile.sidecar-in-ci so that before we set the user to non-root at https://github.com/redhat-appstudio/jvm-build-service/blob/main/Dockerfile.sidecar-in-ci#L18 we create dirs and update permissions so that we no longer the various `chown` calls and use of anyuid scc in our various tasks (i.e. https://github.com/redhat-appstudio/jvm-build-service/blob/3a19938e522f16830cb6c67449ad0bf5117ffc3c/pkg/reconciler/dependencybuild/buildrecipeyaml.go#L253-L254)

The only task not unique to jvm build service is the git-clone task (which is in build-serivce, but I think is basically a copy of the tekton cluster task.  I think it would need to be solved as well long term, but since that effects everyone, probably would need a broader discussion / review.

**Comment by gabemontero** [comment]
some github throttling in the image failure

appstudio bootstrap failure in jvm-build-service-e2e though I just merged a build-service fix that should help there

the sample component maven-goals pod failed, so no artifactbuild / dependencybuild generation:

```
    jvm_build_service_test.go:51: time: 2022-08-02 10:37:07.264819823 +0000 UTC m=+641.164228848: pod logs for container step-mvn-goals in pod sample-component-build-rj2gt-maven-run-pod:  [INFO] Scanning for projects...
        [INFO] 
        [INFO] ------------------< org.example:simple-java-project >-------------------
        [INFO] Building simple-java-project 1.0-SNAPSHOT
        [INFO] --------------------------------[ jar ]---------------------------------
        Downloading from mirror.default: http://localhost:2000/maven2/org/apache/maven/plugins/maven-shade-plugin/3.2.4/maven-shade-plugin-3.2.4.pom
        [INFO] ------------------------------------------------------------------------
        [INFO] BUILD FAILURE
        [INFO] ------------------------------------------------------------------------
        [INFO] Total time:  30.921 s
        [INFO] Finished at: 2022-08-02T10:29:41Z
        [INFO] ------------------------------------------------------------------------
        [ERROR] Plugin org.apache.maven.plugins:maven-shade-plugin:3.2.4 or one of its dependencies could not be resolved: Failed to read artifact descriptor for org.apache.maven.plugins:maven-shade-plugin:jar:3.2.4: Could not transfer artifact org.apache.maven.plugins:maven-shade-plugin:pom:3.2.4 from/to mirror.default (http://localhost:2000/maven2): Transfer failed for http://localhost:2000/maven2/org/apache/maven/plugins/maven-shade-plugin/3.2.4/maven-shade-plugin-3.2.4.pom 500 Internal Server Error -> [Help 1]
        [ERROR] 
        [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
        [ERROR] Re-run Maven using the -X switch to enable full debug logging.
        [ERROR] 
        [ERROR] For more information about the errors and possible solutions, please read the following articles:
        [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginResolutionException
        cat: /workspace/maven-settings/sidecar.log: No such file or directory
```

**Comment by gabemontero** [comment]
hmm ... images failed again ... may not be a flake ... looking again

**Comment by gabemontero** [comment]
Nope - same thing - very weird / bad luck, this PR is not changing what is running elsewhere best as I can tell

/test images

**Comment by stuartwdouglas** [comment]
The sidecar is running out of memory on the image build.

**Comment by gabemontero** [comment]
Last test the operator was complaining about the config map not being available.

But now https://github.com/redhat-appstudio/infra-deployments/pull/537 has merged, so let's try again.

/test jvm-build-service-in-repo-e2e

**Comment by gabemontero** [comment]
OK controller still getting 

```
         1.6595440611397045e+09	ERROR	controller.configmap	Reconciler error	{"reconciler group": "", "reconciler kind": "ConfigMap", "name": "jvm-build-config", "namespace": "jvm-build-service-test-namespace-sls8w", "error": "roles.rbac.authorization.k8s.io \"pipeline-anyuid-role\" not found"}
        sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        	/opt/app-root/src/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266
        sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        	/opt/app-root/src/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227
```

complaining about  the role in particular ... I'll spend cycles now/this afternoon trying to sort out.

File: pkg/reconciler/configmap/configmap.go
Comment (gabemontero) [inline pkg/reconciler/configmap/configmap.go:255]:
```suggestion
			cb.RoleRef = v13.RoleRef{Kind: "ClusterRole", Name: "pipeline-anyuid-role", APIGroup: "rbac.authorization.k8s.io"}
```

**Review by gabemontero — CHANGES_REQUESTED** [review]
I see the problem with the e2e error log I noted ^^ @stuartwdouglas 

You changed `pipeline-anyuid-role` to a `ClusterRole` in https://github.com/redhat-appstudio/jvm-build-service/pull/105/files#diff-aead536da1d8e62277d27b3f65e8a0153ff084e61727d0123d15d881368bfe02 but your code here still cites a namespace scoped `Role`

**Comment by stuartwdouglas** [comment]
@gabemontero jvm-build-service-in-repo-e2e is green, and I have removed the need for privileged containers. I am going to squash this and do some cleanup and it should be good to go.

**Comment by gabemontero** [comment]
> @gabemontero jvm-build-service-in-repo-e2e is green, and I have removed the need for privileged containers. I am going to squash this and do some cleanup and it should be good to go.

awesome on both counts !!

still morning for me, but as of this moment, not going to merge this *yet* because we also need to really get https://github.com/redhat-appstudio/infra-deployments/pull/558 merged to facilitate the e2e work over in e2e-tests, and merging this will interrupt that effort.  And given the amount of non-jvm flakes we are already hitting there, I want to avoid any interrupts I can control.

That said, if we end up at my EOB and that PR is still struggling, I'll merge and we can pick back up on https://github.com/redhat-appstudio/infra-deployments/pull/558 and the various CI testing tomorrow / Monday.

**Comment by gabemontero** [comment]
@stuartwdouglas - can you point out to me where in the PR correlates to your removing the need for privileged containers?

I *think* it relates to earlier forms of the PR where you were assigning the anyuid role, but I don't see that anymore ... is that it?

Otherwise, when I first read of, I thought of https://github.com/redhat-appstudio/jvm-build-service/blob/650846d2011d745b82c72cd59cf1e2398daeddfa/pkg/reconciler/dependencybuild/buildrecipeyaml.go#L137 for example, but of course the use of runAsUser is still there.

thanks

File: Dockerfile.cache-in-ci
Comment (gabemontero) [inline Dockerfile.cache-in-ci:10]: ahh ... I bet this is the no longer need privileged change you are talking about

side bar, is there something special about user 185?  In case you were not aware, when the pod runs in openshift, the actual UID will not be 185.

Comment (gabemontero) [reply Dockerfile.cache-in-ci:10]: I would think this change would allow us to remove `runAsUse: 0` then in https://github.com/redhat-appstudio/jvm-build-service/blob/650846d2011d745b82c72cd59cf1e2398daeddfa/deploy/cache/base/deployment.yaml#L24

did you try that @stuartwdouglas ?

**Comment by gabemontero** [comment]
OK https://github.com/redhat-appstudio/infra-deployments/pull/558 has merged.

That said, I'm now curious about https://github.com/redhat-appstudio/jvm-build-service/pull/105#discussion_r938031556 and a) did you try it and it did not work, or b) did not try it, but are willing to do so in this PR

So I'll defer hitting the merge button @stuartwdouglas to you, where either 
- you merge because the answer to a) was "it did not work" ... I can look into why on Monday unless you already know
- the answer to b) is you want to defer that change to a follow up PR
- you decide to make the change to remove `runAsUser: 0` and then you merge this after CI passes with that change

File: Dockerfile.cache-in-ci
Comment (stuartwdouglas) [reply Dockerfile.cache-in-ci:10]: User 185 comes from the base image, that file is only still there to allow infra deployments to work in the short term, as now the cache is workspace scoped.

**Comment by stuartwdouglas** [comment]
@gabemontero the cache deployment yaml is a transitional thing, once this is merged and infra deployments is updated it will be deleted, which is why I did not look at fixing the user.

I also looked removing it from the tasks but there seems to be an issue with the git clone task not specifying a user and the Java build ones do, which results in the workspace not being writable.

**Comment by gabemontero** [comment]
> @gabemontero the cache deployment yaml is a transitional thing, once this is merged and infra deployments is updated it will be deleted, which is why I did not look at fixing the user.
> 
> I also looked removing it from the tasks but there seems to be an issue with the git clone task not specifying a user and the Java build ones do, which results in the workspace not being writable.

sounds good cache deployment and yeah forgot about the git clone (I remember having to make and update for that now) ... would think that is a app studio wide issue, though I'm not sure if I've seen the need for that in the other build tasks.  I'll revisit that next week.

So yeah merge away at your leisure @stuartwdouglas

**Pull request merged**
